docs/sglang_multiturn/multiturn.rst

```diff
@@ -40,6 +40,82 @@ Finally, set the ``tools_config_file`` in your rollout config:
 
 This allows integration of customized tool behaviors during actor rollout steps. 
 
+Multi-turn Tokenization
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Tokenizing multi-turn rollouts poses a challenge: after applying the chat template and tokenizing the full message list, it’s hard to identify which tokens belong to assistant messages. Since the token list is flat, it lacks direct alignment with the message roles.
+
+To address this, we adopt a **delta-based tokenization** strategy. Each time the LLM generates a new message, we:
+
+1. Apply the chat template to all prior messages (`messages[:i]`).
+2. Apply the chat template again including the latest message (`messages[:i+1]`).
+3. Tokenize only the *delta* between these two serialized message strings.
+
+This ensures that only tokens generated by the assistant are included in the loss mask.
+
+.. code-block:: python
+
+    # Exclude the assistant prompt (e.g., "<|im_start|>assistant") from the loss by setting add_generation_prompt=True
+    prev = tokenizer.apply_chat_template(messages[:i], add_generation_prompt=True, tokenize=False)
+    curr = tokenizer.apply_chat_template(messages[:i+1], add_generation_prompt=False, tokenize=False)
+    token_ids += tokenizer.encode(curr[len(prev):], add_special_tokens=False)
+    loss_mask += [1] * len(token_ids)  # Mask only the new assistant tokens
+
+While we’ve validated this produces consistent results with full message tokenization, future models' chat template could break compatibility. To guard against silent inconsistencies, we compare the delta-based tokenization with full-tokenization results by default at the end of each rollout.
+
+If you see the following warning, enable `INFO` log level to inspect the mismatched outputs:
+
+.. code-block::
+
+    Inconsistent training and inference tokenization detected. This may lead to unexpected behavior during training. Please review your chat template to determine if this is intentional. For more information, refer to the multiturn README.md.
+
+If the discrepancy is expected, you can disable the sanity check via:
+
+``actor_rollout_ref.rollout.multi_turn.enable_tokenization_sanity_check=False``
+
+Special Cases
+^^^^^^^^^^^^^
+
+Some models (e.g., Qwen/QwQ-32B and Qwen3 series) remove internal reasoning content during chat template rendering. As a result, the message content can vary across turns, making the delta-based tokenization inaccurate.
+
+To handle this, we fall back to a **fixed base conversation** containing only a single system and user message. Since this base doesn’t include assistant messages or reasoning content, it remains consistent across turns.
```

Comment:
Since this base doesn’t include assistant messages or reasoning content, it remains consistent across turns. 

We should emphasize the differences between the turns managed by the last assistant turn (QwQ-32B) and those managed by the assistant turn after the user's last input (Qwen3), as these discrepancies lead to trajectory variations. Clarifying this will make the current implementation easier to understand for future users and developers.